### PR TITLE
🔀 add TwyperFlip

### PR DESCRIPTION
**Pull request by yash-k9**
Twyper Flip takes in three extra parameters, two composables, one for front and back and modifier for the box. Animations are added in the existing CardController.

```@Composable
inline fun <reified T> TwyperFlip(
    items: List<T>,
    onItemRemoved: (T, SwipedOutDirection) -> Unit,
    onEmpty: () -> Unit = {},
    twyperController: TwyperController = rememberTwyperController(),
    stackCount: Int = 2,
    paddingBetweenCards: Float = 40f,
    modifier: Modifier = Modifier,
    cardModifier: () -> Modifier,
    crossinline front: @Composable (T) -> Unit,
    crossinline back: @Composable (T) -> Unit
) 
```
https://user-images.githubusercontent.com/59912430/155768674-90167ea2-bd04-456f-a5b1-295cf814d920.mp4

### CODE REVIEW

File: sample/src/main/java/com/github/theapache64/twyper/Util.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package com.github.theapache64.twyper
+
+import androidx.compose.ui.graphics.Color
+
+private val rgbRange = 0..255
+
+object Util {
+    fun randomColor(): Color {
```
Comment (theapache64) [inline sample/src/main/java/com/github/theapache64/twyper/Util.kt]: Looks unused. Do we need this method?

File: twyper/src/main/java/com/github/theapache64/twyper/CardController.kt [hunk 2]
```diff
@@ -21,6 +19,10 @@ interface CardController {
     val cardX: Float
     val cardY: Float
     val rotation: Float
+    val cardFlipRotation: Float
+    val cardFrontAlpha: Float
+    val cardBackAlpha: Float
+    var state: FlipState
     var swipedOutDirection: SwipedOutDirection?
```
Comment (theapache64) [inline twyper/src/main/java/com/github/theapache64/twyper/CardController.kt]: I think this should be inside a separate class. Consider create a new class by subclassing `CardControllerImpl`. The problem is, here we are considering all cards as flippable. But that's not true. Only cards that are inside TwyperFlip is flippable.

File: twyper/src/main/java/com/github/theapache64/twyper/CardController.kt [hunk 3]
```diff
@@ -31,8 +33,12 @@ interface CardController {
      * To check if the card gone out of screen bound
      */
     fun isCardOut(): Boolean
+    fun isCardRotated(): Boolean
+    fun setCardFlipState()
     fun swipeRight()
     fun swipeLeft()
+    fun flipToFront()
+    fun flipToBack()
 }
```
Comment (theapache64) [inline twyper/src/main/java/com/github/theapache64/twyper/CardController.kt]: Same as above

File: twyper/src/main/java/com/github/theapache64/twyper/CardController.kt [hunk 4]
```diff
@@ -44,18 +50,31 @@ fun rememberCardController(): CardController {
     return remember {
         val swipeX = Animatable(0f)
         val swipeY = Animatable(0f)
-        CardControllerImpl(swipeX, swipeY, scope, screenWidth)
+        val flipRotation = Animatable(0f)
+        val frontAlpha = Animatable(1f)
+        val backAlpha = Animatable(0f)
+        val cardState = FlipState.FRONT
+        CardControllerImpl(swipeX, swipeY, flipRotation, frontAlpha, backAlpha, cardState, scope, screenWidth)
     }
```
Comment (theapache64) [inline twyper/src/main/java/com/github/theapache64/twyper/CardController.kt]: Consider create a new method. Something like `rememberFlipCardController`

File: twyper/src/main/java/com/github/theapache64/twyper/CardController.kt [hunk 5]
```diff
@@ -44,18 +50,31 @@ fun rememberCardController(): CardController {
...
+    private val screenWidth: Float
 ) : CardController {
     companion object {
-        private const val SWIPE_DURATION_IN_MILLIS = 450
+        private const val SWIPE_DURATION_IN_MILLIS = 400
+        private const val FLIP_ROTATION_DURATION_IN_MILLIS = 500
+        private const val CARD_FRONT_ALPHA_DURATION_IN_MILLIS = 400
+        private const val CARD_REVERSE_ALPHA_DURATION_IN_MILLIS = 200
+        private const val INITIAL_FLIP_ANGLE = 0f
+        private const val FINAL_FLIP_ANGLE = 180f
```
Comment (theapache64) [inline twyper/src/main/java/com/github/theapache64/twyper/CardController.kt]: Same as above. This should be inside `FlipCardController`

File: twyper/src/main/java/com/github/theapache64/twyper/CardController.kt [hunk 6]
```diff
@@ -64,6 +83,17 @@ class CardControllerImpl(
...
+        get() = flip.value
+
+    override val cardFrontAlpha: Float
+        get() = frontAlpha.value
+
+    override val cardBackAlpha: Float
+        get() = backAlpha.value
+
+    override var state: FlipState = cardState
+
```
Comment (theapache64) [inline twyper/src/main/java/com/github/theapache64/twyper/CardController.kt]: Same as above.

File: twyper/src/main/java/com/github/theapache64/twyper/CardController.kt [hunk 7]
```diff
@@ -105,6 +135,10 @@ class CardControllerImpl(
         return abs(swipeX.value) == screenWidth
     }
 
+    override fun isCardRotated(): Boolean {
+        return cardState == FlipState.BACK
+    }
+
```
Comment (theapache64) [inline twyper/src/main/java/com/github/theapache64/twyper/CardController.kt]: Same as above

File: twyper/src/main/java/com/github/theapache64/twyper/CardController.kt [hunk 8]
```diff
@@ -118,4 +152,30 @@ class CardControllerImpl(
...
+    override fun flipToBack() {
+        scope.launch {
+            cardState = FlipState.BACK
+            launch { flip.animateTo(FINAL_FLIP_ANGLE, tween(FLIP_ROTATION_DURATION_IN_MILLIS)) }
+            launch { frontAlpha.animateTo(0f, tween(CARD_REVERSE_ALPHA_DURATION_IN_MILLIS)) }
+            launch { backAlpha.animateTo(1f, tween(CARD_FRONT_ALPHA_DURATION_IN_MILLIS)) }
+        }
+    }
```
Comment (theapache64) [inline twyper/src/main/java/com/github/theapache64/twyper/CardController.kt]: Same as above

**Review by theapache64 — CHANGES_REQUESTED** [review]

File: sample/src/main/java/com/github/theapache64/twyper/Util.kt (hunk 1 above)
Comment (yash-k9) [reply sample/src/main/java/com/github/theapache64/twyper/Util.kt]: In the Flip version, we are constructing the modifier outside the Composable context, so was not able to use the existing rememberRandomColor() method. Created this method to generate the colors, Shall I modify or place it elsewhere?

File: twyper/src/main/java/com/github/theapache64/twyper/TwyperController.kt [hunk 9]
```diff
@@ -32,4 +33,8 @@ class TwyperControllerImpl : TwyperController {
     override fun swipeLeft() {
         currentCardController?.swipeLeft()
     }
+
```
Comment (yash-k9) [inline twyper/src/main/java/com/github/theapache64/twyper/TwyperController.kt]: Can I create separate controller for the FlipCard and separate the classes to packages, like cardControllers, TwyperControllers?

File: twyper/src/main/java/com/github/theapache64/twyper/CardController.kt (hunk 2 above)
Comment (yash-k9) [reply twyper/src/main/java/com/github/theapache64/twyper/CardController.kt]: Yes, Agreed! 👍

File: sample/src/main/java/com/github/theapache64/twyper/Util.kt (hunk 1 above)
Comment (theapache64) [reply sample/src/main/java/com/github/theapache64/twyper/Util.kt]: I think you can remove this method, and then extract out the color generation logic from the `rememberRandomColor()`. Then you'll be able to use it from both composable and non-composable context.

File: sample/src/main/java/com/github/theapache64/twyper/RandomColor.kt [hunk 10]
```diff
@@ -18,7 +19,13 @@ fun rememberRandomColor(): Color {
             restore = { colorLong -> Color(colorLong.toULong()) }
         )
     ) {
-        Color(
+       randomColor()
```
Comment (yash-k9) [inline sample/src/main/java/com/github/theapache64/twyper/RandomColor.kt:22]: I extracted the color generation logic here, Did I get this right?

Comment (theapache64) [reply sample/src/main/java/com/github/theapache64/twyper/RandomColor.kt:22]: Yeah. this is fine.

**Review by theapache64 — APPROVED** [review]

File: twyper/src/main/java/com/github/theapache64/twyper/TwyperController.kt (hunk 9 above)
Comment (theapache64) [reply twyper/src/main/java/com/github/theapache64/twyper/TwyperController.kt]: I think this is fine for now. We'll refactor the code later. I'll raise a separate PR for that later.

**Pull request merged**